### PR TITLE
DROOLS-4484 / DROOLS-4615: [DMN Designer] Data Types - Refactor the Data Types list to use a drag and drop mechanism / Persist data type list order into the DMN model

### PR DESCRIPTION
**Pull request by karreiro**
See:
- https://issues.jboss.org/browse/DROOLS-4484
- https://issues.jboss.org/browse/DROOLS-4615

Since this is a big PR, I've split it into small meaningful commits.

Another approach for reviewing it is to start with the `DNDDataTypesHandler` class (every time that a user drops an element, this handler is called).

Please let me know if you have any question :-)

### CODE REVIEW

**Comment by danielzhe** [comment]
I found some issues. :(
But I think both are related and probably when you fix one the other one will be fixed too:
**ISSUE 1:**
issue1

**ISSUE 2:**
issue2

**Comment by karreiro** [comment]
Thank you, @danielzhe :-) Working on fixes for these issues right now.

**Comment by karreiro** [comment]
@danielzhe I've fixed both issues in a separated commit.

**Issue 1 (fixed):**
issue1

**Issue 2 (fixed):**
issue2

**Comment by jomarko** [comment]
### Manual Review
#### Issue 1
There is a problem in editing entries once drag and drop was in action.
https://issues.jboss.org/secure/attachment/12456751/issue1-drag-edit.webm

**Comment by jomarko** [comment]
### Manual Review
#### Issue 2
If user start an edit mode with a mouse, then the *Ctrl+S* is not working for saving the changes.
I checked 7.29.0-SNAPSHOT where this issue is not.

**Comment by jomarko** [comment]
### Manual Review
#### Issue 3
If user click on the drag icon, all items are removed.
https://issues.jboss.org/secure/attachment/12456754/issue3-click-hide.webm

**Comment by jomarko** [comment]
### Manual review
#### Issue 4
Included data types can be edited
Screenshot from 2019-10-22 11-39-26

**Comment by jomarko** [comment]
### Manual Review
#### Issue 5
There is some issue when user reorder structures, notice in the video the indicator of the second structure is lost.
https://issues.jboss.org/secure/attachment/12456755/issue5-reorder-structure.webm

**Comment by jomarko** [comment]
### Manual Review
#### Issue 6
When user drag a data type entry that is used in some structure entry, then it is lost in that structure.
https://issues.jboss.org/secure/attachment/12456756/issue6-drag-used-type.webm

**Comment by jomarko** [comment]
Done with manual review, I postpone code review due to found issues.

**Comment by karreiro** [comment]
Thank you, @jomarko :-) Working on fixes for these issues right now.

**Comment by karreiro** [comment]
@jomarko I've fixed the issues above into separated commits. The PR is ready for the re-review.

Regarding the read-only data types. I'm keeping them collapsed for now... I don't clearly know how to handle this kind of Data Type in the drag and drop context. So, I've raised this JIRA https://issues.jboss.org/browse/DROOLS-4677

**Comment by karreiro** [comment]
@jomarko I've fixed the issues above into separated commits.

Regarding the read-only data types. I'm keeping them collapsed for now... we don't clearly know how to handle this kind of Data Type in the drag and drop context. So, I've raised this JIRA https://issues.jboss.org/browse/DROOLS-4677.

The PR is ready for the re-review :-)

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/listview/draganddrop/DNDDataTypesHandler.java [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+
+            final DNDContext dndContext = makeDndContext(currentElement, hoverElement);
+            final Optional<DataType> current = dndContext.getCurrentDataType();
+            final Optional<DataType> reference = dndContext.getReference();
+
+            if (current.isPresent() && reference.isPresent()) {
+                shiftCurrentByReference(current.get(), reference.get(), dndContext.getStrategy());
+            }
+        } catch (final Exception e) {
+            logError("Drag-n-Drop error. Check '" + DNDDataTypesHandler.class.getSimpleName() + "'.");
```
[the remark this answers is not in the document]
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/listview/draganddrop/DNDDataTypesHandler.java]: Fixed!

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/listview/draganddrop/DNDDataTypesHandler.java [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+                    final String errorMessage = "'DNDDataTypesHandler' must be initialized with a 'DNDListComponent' instance.";
+                    return new UnsupportedOperationException(errorMessage);
+                });
+    }
+
+    void logError(final String message) {
+        DomGlobal.console.error(message);
+    }
+
+    class DNDContext {
```
[the remark this answers is not in the document]
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/listview/draganddrop/DNDDataTypesHandler.java]: Fixed!

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/listview/draganddrop/DNDListComponent.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+                    }
+                });
+    }
+
+    private void consolidateHierarchicalLevel() {
+        view.consolidateHierarchicalLevel();
+    }
+
+    public interface View extends UberElemental<DNDListComponent>,
+                                  IsElement {
```
[the remark this answers is not in the document]
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/listview/draganddrop/DNDListComponent.java]: Thank you :-)

**Comment by karreiro** [comment]
@jomarko @lclay2 Changes applied. Thank you for the review.

**Comment by lclay2** [comment]
@karreiro - Another future item we'd like to address is keyboard accessibility. We'll be sure to create a JIRA for tracking purposes after the current release, in the meanwhile we can share this information: https://www.patternfly.org/pattern-library/forms-and-controls/drag-and-drop/#design

**Comment by karreiro** [comment]
@jomarko @danielzhe just friendly reminder. Please take a look when you have a moment :-)

**Comment by danielzhe** [comment]
@karreiro Very good. I loved how I can drag an item to the right and the top item automatically becomes a structure and the item I just dragged becomes a nested field.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/listview/draganddrop/DNDListComponent.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jomarko) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/listview/draganddrop/DNDListComponent.java]: However it is duplicated here, `UberElemental` already extend it.

**Comment by jomarko** [comment]
New reported issue 1
https://issues.jboss.org/browse/DROOLS-4710

**Comment by jomarko** [comment]
New reported issue 2
https://issues.jboss.org/browse/DROOLS-4711

**Comment by jomarko** [comment]
New reported issue 3
https://issues.jboss.org/browse/DROOLS-4712

**Comment by jomarko** [comment]
New reported issue 4
https://issues.jboss.org/browse/DROOLS-4713

**Comment by jomarko** [comment]
New reported issue 5
https://issues.jboss.org/browse/DROOLS-4714

**Comment by jomarko** [comment]
New reported issue 6
https://issues.jboss.org/browse/DROOLS-4715

**Comment by manstis** [comment]
@jomarko I `UberElemental` extends `org.jboss.errai.common.client.api.elemental2.IsElement` and IIRC supports use of the Elemental2 UI library. The other is `org.jboss.errai.ui.client.local.api.IsElement` and supports the use of `errai-ui` HTML element based widgets. The move to Elemental2 is covered by https://issues.jboss.org/browse/DROOLS-4329 (part of https://issues.jboss.org/browse/DROOLS-2974.. I've spoken to Edson about out (DMN) technical debt (with regard to outstanding kogito work and the growing backlog of open JIRA like these 2 I've just linked) as it's becoming an issue IMO and we need to spend as much time fixing some of the older items as we do doing new development....

**Comment by karreiro** [comment]
Merging. Since, I've agreed with @jomarko to make the IsElement-related change in a separate PR.

**Pull request merged**